sql/sql_class.cc

```diff
@@ -895,13 +896,44 @@ void THD::cleanup_connection(void) {
 #endif
 }
 
-void THD::set_connection_certificate(std::string const &cert) {
-  DBUG_ASSERT(m_connection_certificate.empty());
-  m_connection_certificate = cert;
-}
+std::string THD::extract_peer_certificate_info(const THD *thd, bool printable) {
+  if (!thd->has_net_vio_ssl_arg()) return {};
+
+  auto ssl = static_cast<SSL *>(thd->get_net_vio_ssl_arg());
+
+  // extract user cert ref from the thread
+  auto cert_deleter = [](X509 *cert) {
+    if (cert != nullptr) X509_free(cert);
+  };
+  using x509_ptr = std::unique_ptr<X509, decltype(cert_deleter)>;
+
+  x509_ptr cert{SSL_get_peer_certificate(ssl), cert_deleter};
+  if (!cert) return {};
+
+  // Create new X509 buffer abstraction
+  auto bio_deleter = [](BIO *bio) {
+    if (bio != nullptr) BIO_free(bio);
+  };
+  using bio_ptr = std::unique_ptr<BIO, decltype(bio_deleter)>;
+
+  bio_ptr bio{BIO_new(BIO_s_mem()), bio_deleter};
+  if (!bio) return {};
+
+  // Print the certificate to the buffer
+  int print_result = 0;
+  if (printable)
+    print_result = X509_print(bio.get(), cert.get());
+  else
+    print_result = PEM_write_bio_X509(bio.get(), cert.get());
+  if (print_result != 1) return {};
+
+  // decouple buffer and close bio object
```

Comment:
Are you sure this comment is true?
The bio object is closed when the unique_ptr is destructed, at the end of the function. And if it decouples it, and it doesn't free the memory, we are leaking memory - but based on the documentation, I don't think BIO_get_mem_ptr does that.

Reply:
Oh, this is just an outdated comment. There is no decoupling at all. Removing.